src/lib/server/session.cpp

```diff
@@ -0,0 +1,205 @@
+    case PostgresMessageType::BindCommand: {
+      _sync_send_after_error = false;
+      _handle_bind_command();
+      break;
+    }
+    case PostgresMessageType::DescribeCommand: {
+      // The contents of this packet are not used for further processing. The actual "describe" happens after
+      // executing the PQP.
+      _postgres_protocol_handler->read_describe_packet();
+      break;
+    }
+    case PostgresMessageType::ExecuteCommand: {
+      _handle_execute();
+      break;
+    }
+    default:
+      Fail("Unknown packet type");
+  }
+}
+
+void Session::_handle_simple_query() {
+  const auto& query = _postgres_protocol_handler->read_query_packet();
+
+  // A simple query command invalidates unnamed portals
+  _portals.erase("");
+
+  const auto execution_information = QueryHandler::execute_pipeline(query, _send_execution_info);
+
+  if (!execution_information.error.empty()) {
+    _postgres_protocol_handler->send_error_message(execution_information.error);
+  } else {
+    uint64_t row_count = 0;
+    // If there is no result table, e.g. after an INSERT command, we cannot send row data. Otherwise, the result table
+    // of the last statement will be send back.
+    if (execution_information.result_table) {
+      ResultSerializer::send_table_description(execution_information.result_table, _postgres_protocol_handler);
+      ResultSerializer::send_query_response(execution_information.result_table, _postgres_protocol_handler);
+      row_count = execution_information.result_table->row_count();
+    }
+    if (_send_execution_info == SendExecutionInfo::Yes) {
+      _postgres_protocol_handler->send_execution_info(execution_information.pipeline_metrics);
+    }
+    _postgres_protocol_handler->send_command_complete(
+        ResultSerializer::build_command_complete_message(execution_information.root_operator, row_count));
+  }
+  _postgres_protocol_handler->send_ready_for_query();
+}
+
+void Session::_handle_parse_command() {
+  const auto [statement_name, query] = _postgres_protocol_handler->read_parse_packet();
+  QueryHandler::setup_prepared_plan(statement_name, query);
+
+  _postgres_protocol_handler->send_status_message(PostgresMessageType::ParseComplete);
+
+  // Ready for query + flush will be done after reading sync message
+}
+
+void Session::_handle_bind_command() {
+  const auto parameters = _postgres_protocol_handler->read_bind_packet();
+
+  // Named portals must be explicitly closed before they can be redefined by another Bind message,
+  // but this is not required for the unnamed portal.
+  // https://www.postgresql.org/docs/10/static/protocol-flow.html
+  auto portal_it = _portals.find(parameters.portal);
+  if (portal_it != _portals.end()) {
+    AssertInput(parameters.portal.empty(), "Named portals must be explicitly closed before they can be redefined.");
+    _portals.erase(portal_it);
+  }
+
+  // Since describe and execute packet usually arrive together, we still have to handle the execute packet. Therefore,
```

Comment:
Are you sure you mean „describe“ here? Any chance this should read „bind“? Otherwise I am not fully understanding this.

Reply:
Yes, "describe" is wrong. Must be "bind".